### PR TITLE
Approval voting failsafe

### LINKED ISSUE #2561: Governance fail-safe for approval checking
If there is one parablock that doesn't get approved by the approval checking process, for some reason, then the finality of the relay chain will halt and will not recover. There should be a governance fail-safe to manually approve certain relay-chain blocks and get finality going again.

### PR DESCRIPTION
**Pull request by rphmeier**
Closes #2561 

This adds:
  * A `ConsensusLog` digest ported from #2593 with an extra variant for `ForceApprove`
  * A utility `Call` in the `initializer` for a custom origin to dispatch `ForceApprove` digests. This may be e.g. `MoreThanHalfOfCouncil`
  * Code in the approval voting subsystem to monitor for the `ForceApprove` digest and alter on-disk metadata so the blocks will be voted on within GRANDPA.

### CODE REVIEW

File: primitives/src/v1.rs [hunk 1]
```diff
@@ -1049,6 +1049,43 @@ pub struct AbridgedHrmpChannel {
...
+	ParaScheduleUpgradeCode(Id, Hash, BlockNumber),
+	/// Governance requests to auto-approve every candidate included up to the given block
+	/// number in the current chain, inclusive.
+	#[codec(index = 3)]
+	ForceApprove(BlockNumber),
+}
+
+impl ConsensusLog {
+	/// Attempt to convert a reference to a generic digest item into a consensus log.
+	pub fn from_digest_item_ref<H>(digest_item: &runtime_primitives::DigestItem<H>)
```
Comment (bkchr) [inline primitives/src/v1.rs:1072]:
```suggestion
	pub fn from_digest_item<H>(digest_item: &runtime_primitives::DigestItem<H>)
```

File: node/core/approval-voting/src/import.rs [hunk 2]
```diff
@@ -494,13 +495,31 @@ async fn imported_block_info(
 		}
 	};
 
+	let force_approve =
+		block_header.digest.convert_first(|l| match ConsensusLog::from_digest_item_ref(l) {
```
Comment (bkchr) [inline node/core/approval-voting/src/import.rs:499]:
```suggestion
		block_header.digest.convert_first(|l| match ConsensusLog::from_digest_item(l) {
```

File: node/core/approval-voting/src/approval_db/v1/mod.rs [hunk 3]
```diff
@@ -439,6 +441,49 @@ pub(crate) fn add_block_entry(
...
+
+	let mut tx = Transaction::default();
+
+	// iterate back to the `up_to` block, and then iterate backwards until all blocks
+	// are updated.
+	loop {
+		let mut entry = match load_block_entry(store, &cur_hash)? {
+			None => break,
+			Some(e) => e,
+		};
```
Comment (bkchr) [inline node/core/approval-voting/src/approval_db/v1/mod.rs:467]:
```suggestion
	while let Some(mut entry) = load_block_entry(store, &cur_hash)? {
```

**Review by bkchr — APPROVED** [review]

**Comment by rphmeier** [comment]
bot merge

**Review by andresilva — APPROVED** [review]

**Comment by rphmeier** [comment]
bot merge

**Comment by rphmeier** [comment]
bot merge

**Comment by rphmeier** [comment]
got merge

**Comment by rphmeier** [comment]
bot merge

**Pull request merged**